### ISSUE TITLE
[Ray Data] Introduce collate_fn argument in Dataset.iter_torch_batches

### ISSUE THREAD

**Issue by woshiyyya** [ray-project/ray#32224]
### Description

Currently, `iter_torch_batches()` provides the capability of automatic batching, just like what torch.utils.data.dataloader does by default [[link](https://pytorch.org/docs/stable/data.html#automatic-batching-default)].

However, users generally have more complicated input schema and require more flexible batching function. In PyTorch, a common solution is to pass a custom batching function in `collate_fn` argument to the dataloader. (In more than 80% of the training scripts I wrote, I define my own `collate_fn` instead of using the default one.)

However, we have no support on this feature for `iter_torch_batches` yet. Instead, the users have to use a lower-level API `iter_batches` and call the custom batching function separately. I think it would be great to align with PyTorch and have this `collate_fn` argument in `iter_torch_batches`.

```python
def make_batch(rows):
    X = ...
    y = ...
    return X, y

# Pytorch example:
for X, y in DataLoader(dataset, batch_size=10, collate_fn=make_batch):
    X = X.to(device)
    y = y.to(device)
    ...

# Current solution:
for batch in ds.iter_batches(batch_size=10):
    X, y = make_batch(batch)
    X = X.to(device)
    y = y.to(device)
    ...

# An ideal solution:
for X, y in ds.iter_torch_batches(batch_size=10, device=device, collate_fn=make_batch):
    ...
```

### Use case
As far as I know, in many NLP tasks, users requires more complex batching function, some of which cannot be done before model training (e.g. `map_batches` or other per-instance preprocessing function)

Take sentence padding as an example, consider we are training a text classification model, where each training case contains a sentence of variable length. The tokenizer will parse and converted the sentence into a list of token ids. Later, we have to align and pack multiple variable length 1d tensors in a batch into a single 2d tensor on the fly.

For example, we have a batch of 3 cases. Each `x` correspond to a token id in the tokenizers' vocabulary:
```
[x, x, x]
[x, x, x, x, x]
[x, x]
```

Users have to identify the *max_sent_length* among the sequences in the batch, and add paddings for the shorter ones.
```
[[x, x, x, 0, 0],
[x, x, x, x, x],
[x, x, 0, 0, 0]]
```

As we can see, the value of *max_sent_length* can only be calculated after we get a batch. Since the tensor is of variable length, the default automatic batching won't work. That's why we need the `collate_fn`:

```python
def padding_collate_fn(batch):
    max_sent_length = max(len(seq) for seq in batch)
    for seq in batch:
       num_paddings = max_sent_length - len(seq)
       seq += [PAD] * num_paddings
   return torch.FloatTensor(batch)
```

_No response_

**Comment by richardliaw** [comment]
I think we should implement this because this would make the transition from pytorch much more smoother. 

This was a point of confusion from both @woshiyyya (ML team dogfooding) and a customer that @Yard1 has been supporting.

Any thoughts @amogkam @c21 @bveeramani?

**Comment by amogkam** [comment]
+1. Documentation on moving from Torch dataloader to ray datasets has come up multiple times before as well.

**Comment by bveeramani** [comment]
+1 You'd also use `collate_fn` for object detection https://github.com/pytorch/vision/blob/87ec8048085a5d892e78e60cd1fb0b4e37219c1e/references/detection/utils.py#L203-L204

**Comment by clarkzinzow** [comment]
+1 to adding an optional `collate_fn` to the `iter_torch_batches()` API. For the TensorFlow integration, I don't think that we'd want to add such a `collate_fn` since it doesn't align with `tf.data`: the canonical custom batching solution is to either customize the `tf.data.Dataset` construction (i.e. with the `from_generator()` API), which they could do with the `.iter_tf_batches()` API, or use `.apply()`, `.map()`, `.unbatch()`, `.rebatch()`, etc., after receiving an already-batched `tf.data.Dataset`.

Sentence padding for NLP is a great use case example! We added ragged tensor support (or in Torch parlance, nested tensor support) with this use case in mind (along with object detection @bveeramani), but given that Torch's nested tensor support is still experimental and model support being lackluster, we've found padding to continue to be the typical solution.

Also, if we're wanting a nice `collate_fn` transition from the Torch `DataLoader` to Ray Datasets, then we'll want to take a look at the `collate_fn` semantics when automatic batching is enabled/disabled (our equivalent of `batch_size` being set or not), and whether we want to follow similar semantics. https://pytorch.org/docs/stable/data.html#working-with-collate-fn
